### ISSUE TITLE
[Question] spiflash receives incorrect hash when TLUL is sent over FMC

### ISSUE THREAD

**Issue by jackieke724** [lowRISC/opentitan#6829]
My system is shown below. Based on the original earlgrey design, I added another layer of xbar after xbar_peri, and simply moved the u_uart, u_spi_device, u_gpio and u_usbdev to another FPGA. The two FPGA boards are connected through FMC. The custom FMC serialization and deserialization modules are running at 10 MHz with 8-bit data lines.
![image](https://user-images.githubusercontent.com/15860703/120784740-fef50400-c55e-11eb-8628-a944abdc7fa4.png)

Currently, I think only uart is working (figure below). I see bootrom code output whenever I press reset on VC707. spiflash is not able to upload "hello_world".
![image](https://user-images.githubusercontent.com/15860703/120785769-0832a080-c560-11eb-90e5-35dc7463c53d.png)

In my setup, the latency to access the four peripherals will be larger than before. Does this fail the timing required by spiflash? I tried to increase the `hash_*_delay` and `hash_read_timeout_us`, but that did not help.

**Comment by jackieke724** [comment]
For this initial bring-up, I have not moved the `rv_plic` to the peripheral FPGA. Therefore, interrupts are not supported.

Nevertheless, I do not think interrupts are the problem. Boot rom seems to be just polling the number of bytes in the spi_device rx fifo. Could someone please confirm?

**Comment by moidx** [comment]
Hi @jackieke724, you are correct, boot_rom does not use interrupt, and relies on polling the spi device buffer. Based on your screen capture it seems that mask_rom is not receiving any data. This could be due to timing issues.

Assuming there are no timing issues on the FPGA side, you can try to tweak the timing parameters on the host tool side: 
https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/spiflash/ftdi_spi_interface.h;l=52

**Comment by jackieke724** [comment]
Thank you @moidx for your help! I am not sure if tuning SPI frequency could help, but the problem now is boot_rom has stopped polling spi device buffer after some time after reset.

In ILA, I saw pwrmgr_fsm/i_fsm is still in StActive, so Ibex has not entered low power/sleep mode. Do you have any clues why boot_rom has stopped polling?

**Comment by jackieke724** [comment]
What signals/modules do you suggest to check for the status of ibex or where the instruction got stuck? From observation, the time boot_rom stalled after reset was not constant.

**Comment by jackieke724** [comment]
I have finally found the problem... 

Boot_rom was stuck after the system reset by spiflash (code below). It wasn't until I randomly came across #1159 that I got to fully understand how the `DPS0-7` pins were used. Basically, besided IO_RSTN, spiflash can also reset the whole system using `IO_DPS5`. Spiflash's reset is anded with `IO_RSTN` in clkgen. I did not connect this reset with the other FPGA (the left one in my figure). Therefore, my FMC serialization pairs became out-of-sync.
https://github.com/lowRISC/opentitan/blob/e416cb84c1ce6e31b80f335e8b39b349eeddf93f/sw/host/spiflash/ftdi_spi_interface.cc#L46-L60


In addition, I still need to decrease the frequency to compare and read hash in spiflash (code below), due to the added latency of my FMC pairs.
https://github.com/lowRISC/opentitan/blob/f8d9c7e12fd6c9d85ff752fa62eaf1fa284ab07f/sw/host/spiflash/ftdi_spi_interface.h#L39-L44

I changed to the following. The numbers should be tuned for other scenarios.
```
    /** Time to wait between attempts to check the hash in microseconds. */
    int32_t hash_read_delay_us = 100000;

    /** Time to wait between reading the hash over SPI and checking it in
     *  microseconds. */
    int32_t hash_check_delay_us = 100000;

```

Thank you @moidx  for the help!
